### DISCUSSION TITLE
Scheduler is frozen after one month of running.

### DISCUSSION THREAD

**Discussion by danies8** [quartznet/quartznet#2351] [Q&A] [answered]
Hi,
I run my app one month without problems with a lot of files .After one month the scheduler is frozen.
I'm using QUARTZ Version 3.8.0 on net 8.
How can i know when the scheduler is frozen in order to send email to our customers to restart the scheduers.
Any api that can tell me about this issue.
This system in production.
Do you have new version that solved the problem maybee?
There is no report on my logs and no on event viewer.
How can i monitor the scheduler if it stopped working and restart it automatically?
Your quick answer will be appreciated.

**Comment by danies8** [comment]
From your code sample you have this.
Which event I need to use of frozen and what to do?
#region License

/* 

* All content copyright Marko Lahma, unless otherwise indicated. All rights reserved.

* 

* Licensed under the Apache License, Version 2.0 (the "License"); you may not 

* use this file except in compliance with the License. You may obtain a copy 

* of the License at 

* 

*   http://www.apache.org/licenses/LICENSE-2.0 

*   

* Unless required by applicable law or agreed to in writing, software 

* distributed under the License is distributed on an "AS IS" BASIS, WITHOUT 

* WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the 

* License for the specific language governing permissions and limitations 

* under the License.

* 

*/

#endregion
 
namespace Quartz.Listener;
 
/// <summary>

///  A helpful abstract base class for implementors of 

/// <see cref="ISchedulerListener" />.

/// </summary>

/// <remarks>

/// The methods in this class are empty so you only need to override the  

/// subset for the <see cref="ISchedulerListener" /> events you care about.

/// </remarks>

/// <author>Marko Lahma (.NET)</author>

/// <seealso cref="ISchedulerListener" />

public abstract class SchedulerListenerSupport : ISchedulerListener

{

    public virtual ValueTask JobScheduled(

        ITrigger trigger,

        CancellationToken cancellationToken = default)

    {

        return default;

    }
 
    public virtual ValueTask JobUnscheduled(

        TriggerKey triggerKey,

        CancellationToken cancellationToken = default)

    {

        return default;

    }
 
    public virtual ValueTask TriggerFinalized(

        ITrigger trigger,

        CancellationToken cancellationToken = default)

    {

        return default;

    }
 
    public virtual ValueTask TriggersPaused(

        string? triggerGroup,

        CancellationToken cancellationToken = default)

    {

        return default;

    }
 
    public virtual ValueTask TriggerPaused(

        TriggerKey triggerKey,

        CancellationToken cancellationToken = default)

    {

        return default;

    }
 
    public virtual ValueTask TriggersResumed(

        string? triggerGroup,

        CancellationToken cancellationToken = default)

    {

        return default;

    }
 
    public virtual ValueTask TriggerResumed(

        TriggerKey triggerKey,

        CancellationToken cancellationToken = default)

    {

        return default;

    }
 
    public virtual ValueTask JobAdded(

        IJobDetail jobDetail,

        CancellationToken cancellationToken = default)

    {

        return default;

    }
 
    public virtual ValueTask JobDeleted(

        JobKey jobKey,

        CancellationToken cancellationToken = default)

    {

        return default;

    }
 
    public virtual ValueTask JobInterrupted(

        JobKey jobKey,

        CancellationToken cancellationToken = new CancellationToken())

    {

        return default;

    }
 
    public virtual ValueTask JobsPaused(

        string jobGroup,

        CancellationToken cancellationToken = default)

    {

        return default;

    }
 
    public virtual ValueTask JobPaused(

        JobKey jobKey,

        CancellationToken cancellationToken = default)

    {

        return default;

    }
 
    public virtual ValueTask JobsResumed(

        string jobGroup,

        CancellationToken cancellationToken = default)

    {

        return default;

    }
 
    public virtual ValueTask JobResumed(

        JobKey jobKey,

        CancellationToken cancellationToken = default)

    {

        return default;

    }
 
    public virtual ValueTask SchedulerError(

        string msg,

        SchedulerException cause,

        CancellationToken cancellationToken = default)

    {

        return default;

    }
 
    public virtual ValueTask SchedulerInStandbyMode(CancellationToken cancellationToken = default)

    {

        return default;

    }
 
    public virtual ValueTask SchedulerStarted(CancellationToken cancellationToken = default)

    {

        return default;

    }
 
    public virtual ValueTask SchedulerStarting(CancellationToken cancellationToken = default)

    {

        return default;

    }
 
    public virtual ValueTask SchedulerShutdown(CancellationToken cancellationToken = default)

    {

        return default;

    }
 
    public virtual ValueTask SchedulerShuttingdown(CancellationToken cancellationToken = default)

    {

        return default;

    }
 
    public virtual ValueTask SchedulingDataCleared(CancellationToken cancellationToken = default)

    {

        return default;

    }

}

**Reply by danies8** [reply]
@lahma 
Which method I need to override when the sceduler is frozen ?
Can i restart the sceduler if yes how ?

**Comment by danies8** [comment]
@lahma 
This the log that I got when scheduler is "frozen":
Which event I need to override in order to restart the hosted service and the sHceduler.
Your propmt answer will be appreciated.
2024-05-12 09:11:14,032 [.NET TP Worker] DEBUG AutoamtionService.Shared.Helpers.MongoDbHelper - Run the scheduler
2024-05-12 09:11:14,032 [.NET TP Worker] DEBUG AutoamtionService.Shared.Helpers.MongoDbHelper - Config default routes data.
2024-05-12 09:11:14,038 [.NET TP Worker] ERROR Quartz.Core.QuartzScheduler - Error while notifying SchedulerListener of scheduler starting.
System.NullReferenceException: Object reference not set to an instance of an object.
   at Quartz.Core.QuartzScheduler.NotifySchedulerListeners(Func`2 notifier, String action)
2024-05-12 09:11:14,047 [.NET TP Worker] INFO Quartz.Core.QuartzScheduler - Scheduler DefaultQuartzScheduler_$_NON_CLUSTERED started.
2024-05-12 09:11:14,048 [.NET TP Worker] ERROR Quartz.Core.QuartzScheduler - Error while notifying SchedulerListener of startup.
System.NullReferenceException: Object reference not set to an instance of an object.
   at Quartz.Core.QuartzScheduler.NotifySchedulerListeners(Func`2 notifier, String action)
2024-05-12 09:11:14,048 [.NET TP Worker] DEBUG AutoamtionService.Shared.Helpers.MongoDbHelper - Config the scan history scheduler with routes data once.
2024-05-12 09:11:14,051 [.NET TP Worker] ERROR Quartz.Core.QuartzScheduler - Error while notifying SchedulerListener of job addition.
System.NullReferenceException: Object reference not set to an instance of an object.
   at Quartz.Core.QuartzScheduler.NotifySchedulerListeners(Func`2 notifier, String action)
2024-05-12 09:11:14,051 [.NET TP Worker] ERROR Quartz.Core.QuartzScheduler - Error while notifying SchedulerListener of scheduled job. Trigger=ScanHistoryGroupName.ScanHistoryJobName.
System.NullReferenceException: Object reference not set to an instance of an object.
   at Quartz.Core.QuartzScheduler.NotifySchedulerListeners(Func`2 notifier, String action)
2024-05-12 09:11:14,052 [.NET TP Worker] DEBUG AutoamtionService.Shared.Helpers.MongoDbHelper - Run the scan history scheduler
2024-05-12 09:11:14,053 [.NET TP Worker] ERROR Quartz.Core.QuartzScheduler - Error while notifying SchedulerListener of scheduler starting.
System.NullReferenceException: Object reference not set to an instance of an object.
   at Quartz.Core.QuartzScheduler.NotifySchedulerListeners(Func`2 notifier, String action)
2024-05-12 09:11:14,054 [.NET TP Worker] INFO Quartz.Core.QuartzScheduler - Scheduler DefaultQuartzScheduler_$_NON_CLUSTERED started.
2024-05-12 09:11:14,054 [.NET TP Worker] ERROR Quartz.Core.QuartzScheduler - Error while notifying SchedulerListener of startup.
System.NullReferenceException: Object reference not set to an instance of an object.
   at Quartz.Core.QuartzScheduler.NotifySchedulerListeners(Func`2 notifier, String action)
2024-05-12 09:11:14,058 [.NET TP Worker] DEBUG Quartz.Core.QuartzSchedulerThread - Batch acquisition of 1 triggers
2024-05-12 09:11:14,062 [.NET TP Worker] DEBUG Quartz.Simpl.SimpleJobFactory - Producing instance of Job 'DefaultGroup.DefaultJob', class=AutoamtionService.Shared.Helpers.JobSchedulerHelper
2024-05-12 09:11:14,064 [.NET TP Worker] DEBUG Quartz.Core.QuartzSchedulerThread - Batch acquisition of 1 triggers
2024-05-12 09:11:14,064 [.NET TP Worker] DEBUG Quartz.Simpl.SimpleJobFactory - Producing instance of Job 'ScanHistoryGroupName.ScanHistoryJobName', class=AutomationClient.Shared.Helpers.ScanHistoryJobSchedulerHelper
2024-05-12 09:11:14,065 [.NET TP Worker] DEBUG Quartz.Core.QuartzSchedulerThread - Batch acquisition of 1 triggers

**Comment by lahma** [comment]
So you are running officially unsupported MongoDB and you have your own custom listener that is throwing an error. Can't help much more than there's a typo in your `AutoamtionService` 🤷🏻‍♂️ You need to debug and fix your code I guess.

**Comment by danies8** [comment]
@lahma 

Thanks for your quick answer.
1. This my SchedulerListenerHelper  - any wrong with it ?
      _scheduler.ListenerManager.AddSchedulerListener(new SchedulerListenerHelper(_scheduler));
 
 public sealed class SchedulerListenerHelper : SchedulerListenerSupport
  {
      private static AppLogManager _logger = AppLogManager.GetLogger<SchedulerListenerHelper>();
      private readonly IScheduler _scheduler;
      public SchedulerListenerHelper(IScheduler scheduler)
      {
          _scheduler = scheduler;
      }
      public override Task SchedulerError(
     string msg,
     SchedulerException cause,
     CancellationToken cancellationToken = default)
      {
          _logger.Error($"Scheduler Error-{msg}", cause);
          _scheduler.Shutdown();
          if (ServiceHelper.IsWindowServiceRunning(ConstantsHelper.GSAutomationServiceName))
          {
              ServiceHelper.StopService(ConstantsHelper.GSAutomationServiceName);
          }
          return default;
      }
 

  }

2. Why I can't use mongo DB within IJob implementor for my private use?
3. The mongo does not keep data scheduler. I used to create jobs and triggers before schedule them.

**Reply by lahma** [reply]
1. Why would you shutdown the very scheduler that is dispatching you a message in the middle of message dispatch?
2. So if you are not using it as a backing store that would be fine from Quartz's perspective. Just do note that it's not the greatest database.

**Reply by danies8** [reply]
@lahma 
We are searching  two months why the scheduler is "frozen" after uploading half milion files to our website without success.
When it frozen => the service log is stopping. no reports of scheduler.
After reading your samples code I found this code, monitor the scheduler=> SchedulerListenerHelper .
1. Can you please give fixed sniped code, what to do , how to use it effectively ? the scheduler is frozen and i want to close it and it service.
and when user start service again the sceduler will work as expected..

**Reply by lahma** [reply]
Maybe you just need to take process snapshots and examine what is blocking/hanging.

**Reply by danies8** [reply]
@lahma
I would like to ask you about 'process snapshots' is  a tool that i can examine what is blocking/hanging?
I did'nt find in the internet.

**Reply by lahma** [reply]
I think the very first Google result is quite good: https://www.google.fi/search?q=what+is+a+process+snapshot

**Reply by danies8** [reply]
@lahma 
From Microsoft forum:
It may be due to some resource leaks or memory leaks caused by improper resource management. In long-running background tasks, special attention needs to be paid to releasing resources and managing memory to avoid this situation.
Your code may have resource leaks, such as not closing a file in time after opening it, etc. Please make sure to release resources promptly after using them, either using a using statement or manually releasing resources.
 I recommend that you restart background tasks regularly.

Does restart the service once a day will solve the problem ?
Thanks in advance,

**Reply by lahma** [reply]
One option is to run a job that only does some simple logging call or similar check if it will break down or not. If it doesn't, the problem probably is in your code.

**Reply by danies8** [reply, marked as answer]
@lahma 
Thanks for suggestion and your paient.
I found the bug, it was in cron expression in month section, it give only the **current month**.

**Reply by lahma** [reply]
You should always use `TriggerUtils`  and check that your firing schedule will be at least remotely correct.
